{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Profile HMMs\n",
    "\n",
    "---\n",
    "## Learning Objectives\n",
    "\n",
    "1. Apply HMMs to describe PSSMs\n",
    "* Develop model structure for Profile HMM\n",
    "* Build and apply model\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "Today we will be reviewing Profile HMMs in class including a demonstration of how we can implement profile HMMs using our existing framework. \n",
    "\n",
    "This is a diagram of Hidden Markov Model used in HMMER (from the HMMER User Guide by Sean Eddy). The chain of match (M), insert (I), and deletion (D) states can be extended to match the length of the multiple sequence alignment that is used as the training set to produce a model. Individual sequences may then be aligned to the model and scored based on the probability that the model would emit that sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center><img src='./figures/HMM_Diagram.PNG'/></center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "We will be implementing a Profile HMM using the BAR domain discussed in the slides.\n",
    "\n",
    "To accomplish this, we will implement two functions. First, `get_valid_states()` will provide a list of states that meet our heuristic threshold in the model (denoted as *s in the slides). Second, we will implement `build_profileHMM()` that will use our existing HMM class structure (inlcuded here in HMM.py) to develop a model in the above structure.\n",
    "\n",
    "A few caveats: Our HMM implementation requires that all possible emissions and transitions exist in the dictionary. That is, any hidden state must have probabilities of emiting the entire alphabet and the transition matrix must have a probability for every state going to every other state in the model. These probabilites can be set to 0 to create the profile HMM structure, but they must be set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "from data_readers import get_fasta\n",
    "from collections import Counter, defaultdict\n",
    "from HMM import HMM\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_valid_states(fasta_file, threshold=0.7):\n",
    "    ''' \n",
    "    Function to determine which positions in an alignment are valid states in the profile HMM given a threshold\n",
    "    \n",
    "    Args: \n",
    "        fasta_file (str): fasta file containing alignments\n",
    "        threshold (float): the treshold of allowed gap characters (default = 0.7)\n",
    "\n",
    "    Returns:\n",
    "        valid_states (list of bools): list of booleans (True/False) if each position is above the threshold\n",
    "        \n",
    "    Example:\n",
    "        >>> get_valid_states(\"data/BAR.fa\", 0.7) #doctest: +ELLIPSIS\n",
    "        [True, True, True, True, False, ...]\n",
    "    '''\n",
    "    # Returns list of valid states positions\n",
    "\n",
    "    state_counter = None\n",
    "    valid_states = []\n",
    "    \n",
    "    #Estimate the states in the alignment that are match versus indel states\n",
    "    for name, seq in get_fasta(fasta_file):\n",
    "        if state_counter is None:\n",
    "            state_counter = [Counter() for _ in range(0, len(seq))]\n",
    "            \n",
    "        for i, aa in enumerate(seq):\n",
    "            state_counter[i].update(aa)\n",
    "    \n",
    "    \n",
    "    for i, position_counter in enumerate(state_counter):\n",
    "        if position_counter['-']/sum(position_counter.values()) > (1 - threshold):\n",
    "            valid_states.append(False)\n",
    "        else:\n",
    "            valid_states.append(True)\n",
    "\n",
    "    return valid_states"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As mentioned in the slides, we need to update the states for the HMM using the following equations:\n",
    "\n",
    "$a_{kl} = A_{kl} / \\sum_{l'}A_{kl'}$\n",
    "\n",
    "$e_{k}(a) = E_{k}(a) / \\sum_{a'}E_{k}(a')$\n",
    "\n",
    "Where $k$ and $l$ represent state indices, $a_{kl}$ and $e_{k}$ are transition and emission probabilities, respectively, and $A_{kl}$ and $E_{k}$ are the corresponding frequencies.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Train model using the BAR domain data in data/BAR.fa\n",
    "\n",
    "# In order to build our model, we will need to set default paramters in an \n",
    "# initialized HMM using pseudocounts and then update these values with the \n",
    "# information in the fasta file\n",
    "def build_profileHMM(alphabet, valid_states, fasta_file, pseudocount=0.01):\n",
    "    ''' \n",
    "    Function to initialize a Profile HMM structure\n",
    "    \n",
    "    Args: \n",
    "        alphabet (list): alphabet characters for the model\n",
    "        valid_states (list of bools): all positions in the alignments that are in match states\n",
    "        fasta_file (str): fasta file containing alignments\n",
    "        pseudocount (float): value to set as initial probabilities\n",
    "\n",
    "    Returns:\n",
    "        profile_HMM (HMM): HMM object\n",
    "        \n",
    "    Pseudocode:\n",
    "    Initialize full initial, emission, and transition matrix to 0s (our HMM object requires all transitions and emissions are set to at least 0)\n",
    "    Initialize possible emissions and transitons to pseudocount\n",
    "    Calculate probabilities at each position given valid_states and fasta_file\n",
    "        \n",
    "    '''\n",
    "    \n",
    "    #Initialize empty matrices A, E, and I\n",
    "    A = defaultdict(dict)\n",
    "    E = defaultdict(dict)\n",
    "    I = {}\n",
    "    hidden_states = []\n",
    "    \n",
    "    #generate all valid states\n",
    "    #Initialize I0 state as insert at start\n",
    "    hidden_states.append('I0')\n",
    "\n",
    "    #Initialize our states based on the number of match states\n",
    "    for i in range(1, sum(valid_states)+1):\n",
    "        hidden_states.append('I'+str(i))\n",
    "        hidden_states.append('M'+str(i))\n",
    "        hidden_states.append('D'+str(i))\n",
    "    \n",
    "    # Initialize all transitions to 0\n",
    "    for item in hidden_states:\n",
    "        for next_item in hidden_states:\n",
    "            A[item][next_item] = 0\n",
    "\n",
    "            \n",
    "    # Give all valid state transitions a pseudocount\n",
    "    A['I0']['I0'] = pseudocount\n",
    "    \n",
    "    #Initialize our states based on the number of match states\n",
    "    for i in range(1, sum(valid_states)+1):\n",
    "        # Transitions into insert state\n",
    "        A['I'+str(i)]['I'+str(i)] = pseudocount # self\n",
    "        A['D'+str(i)]['I'+str(i)] = pseudocount # from deletion      \n",
    "        A['M'+str(i)]['I'+str(i)] = pseudocount # from match\n",
    "        \n",
    "        # Transitions into deletion state\n",
    "        A['I'+str(i-1)]['D'+str(i)] = pseudocount # from previous insert\n",
    "        if i > 1: #special case for first location so ignore these\n",
    "            A['D'+str(i-1)]['D'+str(i)] = pseudocount # from previous deletion      \n",
    "            A['M'+str(i-1)]['D'+str(i)] = pseudocount # from previous match\n",
    "        \n",
    "        # Transitions into match state\n",
    "        A['I'+str(i-1)]['M'+str(i)] = pseudocount # from previous insert\n",
    "        if i > 1: #special case for first location so ignore these\n",
    "            A['D'+str(i-1)]['M'+str(i)] = pseudocount # from previous deletion      \n",
    "            A['M'+str(i-1)]['M'+str(i)] = pseudocount # from previous match\n",
    "        \n",
    "    # Initialize start state similarly\n",
    "    for item in hidden_states:\n",
    "        I[item] = 0\n",
    "    I['I0'] = pseudocount\n",
    "    I['D1'] = pseudocount\n",
    "    I['M1'] = pseudocount\n",
    "    \n",
    "    # Initialize emissions from Insert and Match states\n",
    "    for i in range(0, sum(valid_states)+1):\n",
    "        for j, aa in enumerate(alphabet):\n",
    "            E['D' + str(i)][aa] = 0\n",
    "            E['I' + str(i)][aa] = pseudocount\n",
    "            if i > 0:\n",
    "                E['M' + str(i)][aa] = pseudocount\n",
    "            else:\n",
    "                E['M' + str(i)][aa] = 0\n",
    "\n",
    "    \n",
    "    # now iterate through the file and train all of the states \n",
    "    for name, seq in get_fasta(fasta_file):\n",
    "        last_state = None\n",
    "        state_number = 1\n",
    "        for i, aa in enumerate(seq):\n",
    "            if i == 0: # first position\n",
    "                if valid_states[i]: # we are in a match or deletion\n",
    "                    if aa == '-': # we are in a deletion\n",
    "                        I['D1'] += 1\n",
    "                        last_state = 'D1'\n",
    "                        state_number += 1\n",
    "                    else: # we are in a match\n",
    "                        I['M1'] += 1\n",
    "                        E['M1'][aa] += 1\n",
    "                        last_state = 'M1'\n",
    "                        state_number += 1\n",
    "                        \n",
    "                else: # We are in an insert or a deletion\n",
    "                    if aa == '-': # we are in a valid location so ignore this\n",
    "                        pass\n",
    "                    else: #we are in an insert\n",
    "                        I['I0'] += 1\n",
    "                        E['I0'][aa] += 1\n",
    "                        last_state = 'I0'\n",
    "                    \n",
    "            else: # same except after the first position\n",
    "                if valid_states[i]: # we are in a match or deletion\n",
    "                    if aa == '-': # we are in a deletion\n",
    "                        A[last_state]['D' + str(state_number)] += 1\n",
    "                        last_state = 'D' + str(state_number)\n",
    "                        state_number += 1\n",
    "                    else: # we are in a match\n",
    "                        A[last_state]['M' + str(state_number)] += 1\n",
    "                        E['M' + str(state_number)][aa] += 1\n",
    "                        last_state = 'M' + str(state_number)\n",
    "                        state_number += 1\n",
    "                        \n",
    "                else: # We are in an insert or a deletion\n",
    "                    if aa == '-': # we are in a valid location so ignore this\n",
    "                        pass\n",
    "                    else: #we are in an insert\n",
    "                        A[last_state]['I' + str(state_number-1)] += 1\n",
    "                        E['I' + str(state_number-1)][aa] += 1\n",
    "                        last_state = 'I' + str(state_number-1)               \n",
    "\n",
    "    # now we normalize to a sum of 1\n",
    "    for items in A:\n",
    "        A_sum = sum(A[items].values())\n",
    "        if A_sum > 0:\n",
    "            for next_item in A[items]:\n",
    "                A[items][next_item] = A[items][next_item] / A_sum\n",
    "\n",
    "    for items in E:\n",
    "        E_sum = sum(E[items].values())\n",
    "        if E_sum > 0:\n",
    "            for next_item in E[items]:\n",
    "                E[items][next_item] = E[items][next_item] / E_sum\n",
    "            \n",
    "    I_sum = sum(I.values())\n",
    "    for items in I:\n",
    "        I[items] = I[items] / I_sum\n",
    "    \n",
    "    return HMM(alphabet, hidden_states, A, E, I)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_states = get_valid_states(\"data/BAR_Short.fa\", 0.5)\n",
    "alphabet = list('GALMFWKQESPVICYHRNDT')\n",
    "profile = build_profileHMM(alphabet, valid_states, \"data/BAR_Short.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exact example from slides\n",
    "sequence = \"TKLDDDFKE\"\n",
    "\n",
    "print (\"Forward:\")\n",
    "f_Px, f_matrix = profile.forward(list(sequence))\n",
    "print (f_Px)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Expected output:\n",
    "Forward:\n",
    "9.605647218365268e-05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(profile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
