{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# eQTL analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Before Class\n",
    "1. Review previous two classes\n",
    "* Review slides on eQTL analsis\n",
    "\n",
    "---\n",
    "## Learning Objectives\n",
    "1. Understand bulk processing of data\n",
    "* Perform eQTL analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Bulk data processing\n",
    "Another way we can predict the effect of variants on expression is through study of the genotypes and not allele-specific expression. For today's class, we have expanded our study to include 100 DNA and RNA samples from diploid individuals from our same small reference genome. Our goal now will be to study if any of the variants in the genome can be associated with the expression of a gene. These variants are said to be eQTLs (expression quantitative trait loci) and help explain variation of expression levels of genes. It is important to note that while the variants are assoiciated with expression changes, they are not necessairly causative of the expression change.\n",
    "\n",
    "The first step of today's class will be to just process all of the RNA and DNA data by aligning it to the genome and converting to bam files. The RNA bam files will need to be indexed as the previous class.\n",
    "\n",
    "Hint: A python loop can be put around the shell commands we've previously used and variables can be used in the shell commands by surrounding them with curly brackets '{}'.\n",
    "For example:\n",
    "```\n",
    "for i in range(10):\n",
    "    !echo {i}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22531 (98.17%) aligned exactly 1 time\n",
      "    421 (1.83%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22547 (98.24%) aligned exactly 1 time\n",
      "    405 (1.76%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22524 (98.14%) aligned exactly 1 time\n",
      "    428 (1.86%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22532 (98.17%) aligned exactly 1 time\n",
      "    420 (1.83%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22527 (98.15%) aligned exactly 1 time\n",
      "    425 (1.85%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22500 (98.03%) aligned exactly 1 time\n",
      "    452 (1.97%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22555 (98.27%) aligned exactly 1 time\n",
      "    397 (1.73%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22537 (98.19%) aligned exactly 1 time\n",
      "    415 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22599 (98.46%) aligned exactly 1 time\n",
      "    353 (1.54%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22523 (98.13%) aligned exactly 1 time\n",
      "    429 (1.87%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22542 (98.21%) aligned exactly 1 time\n",
      "    410 (1.79%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22588 (98.41%) aligned exactly 1 time\n",
      "    364 (1.59%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22542 (98.21%) aligned exactly 1 time\n",
      "    410 (1.79%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22541 (98.21%) aligned exactly 1 time\n",
      "    411 (1.79%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22570 (98.34%) aligned exactly 1 time\n",
      "    382 (1.66%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22551 (98.25%) aligned exactly 1 time\n",
      "    401 (1.75%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22530 (98.16%) aligned exactly 1 time\n",
      "    422 (1.84%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22536 (98.19%) aligned exactly 1 time\n",
      "    416 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22538 (98.20%) aligned exactly 1 time\n",
      "    414 (1.80%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22562 (98.30%) aligned exactly 1 time\n",
      "    390 (1.70%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22541 (98.21%) aligned exactly 1 time\n",
      "    411 (1.79%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22532 (98.17%) aligned exactly 1 time\n",
      "    420 (1.83%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22574 (98.35%) aligned exactly 1 time\n",
      "    378 (1.65%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22556 (98.27%) aligned exactly 1 time\n",
      "    396 (1.73%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22493 (98.00%) aligned exactly 1 time\n",
      "    459 (2.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22538 (98.20%) aligned exactly 1 time\n",
      "    414 (1.80%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22499 (98.03%) aligned exactly 1 time\n",
      "    453 (1.97%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22501 (98.04%) aligned exactly 1 time\n",
      "    451 (1.96%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22544 (98.22%) aligned exactly 1 time\n",
      "    408 (1.78%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22548 (98.24%) aligned exactly 1 time\n",
      "    404 (1.76%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22558 (98.28%) aligned exactly 1 time\n",
      "    394 (1.72%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22507 (98.06%) aligned exactly 1 time\n",
      "    445 (1.94%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22540 (98.20%) aligned exactly 1 time\n",
      "    412 (1.80%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22554 (98.27%) aligned exactly 1 time\n",
      "    398 (1.73%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22536 (98.19%) aligned exactly 1 time\n",
      "    416 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22555 (98.27%) aligned exactly 1 time\n",
      "    397 (1.73%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22529 (98.16%) aligned exactly 1 time\n",
      "    423 (1.84%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22544 (98.22%) aligned exactly 1 time\n",
      "    408 (1.78%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22537 (98.19%) aligned exactly 1 time\n",
      "    415 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22541 (98.21%) aligned exactly 1 time\n",
      "    411 (1.79%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22567 (98.32%) aligned exactly 1 time\n",
      "    385 (1.68%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22539 (98.20%) aligned exactly 1 time\n",
      "    413 (1.80%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22565 (98.31%) aligned exactly 1 time\n",
      "    387 (1.69%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22536 (98.19%) aligned exactly 1 time\n",
      "    416 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22529 (98.16%) aligned exactly 1 time\n",
      "    423 (1.84%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22510 (98.07%) aligned exactly 1 time\n",
      "    442 (1.93%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22536 (98.19%) aligned exactly 1 time\n",
      "    416 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22557 (98.28%) aligned exactly 1 time\n",
      "    395 (1.72%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22527 (98.15%) aligned exactly 1 time\n",
      "    425 (1.85%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22508 (98.07%) aligned exactly 1 time\n",
      "    444 (1.93%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22537 (98.19%) aligned exactly 1 time\n",
      "    415 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22571 (98.34%) aligned exactly 1 time\n",
      "    381 (1.66%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22522 (98.13%) aligned exactly 1 time\n",
      "    430 (1.87%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22572 (98.34%) aligned exactly 1 time\n",
      "    380 (1.66%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22554 (98.27%) aligned exactly 1 time\n",
      "    398 (1.73%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22497 (98.02%) aligned exactly 1 time\n",
      "    455 (1.98%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22515 (98.10%) aligned exactly 1 time\n",
      "    437 (1.90%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22537 (98.19%) aligned exactly 1 time\n",
      "    415 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22556 (98.27%) aligned exactly 1 time\n",
      "    396 (1.73%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22552 (98.26%) aligned exactly 1 time\n",
      "    400 (1.74%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22560 (98.29%) aligned exactly 1 time\n",
      "    392 (1.71%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22554 (98.27%) aligned exactly 1 time\n",
      "    398 (1.73%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22569 (98.33%) aligned exactly 1 time\n",
      "    383 (1.67%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22539 (98.20%) aligned exactly 1 time\n",
      "    413 (1.80%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22501 (98.04%) aligned exactly 1 time\n",
      "    451 (1.96%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22566 (98.32%) aligned exactly 1 time\n",
      "    386 (1.68%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22525 (98.14%) aligned exactly 1 time\n",
      "    427 (1.86%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22555 (98.27%) aligned exactly 1 time\n",
      "    397 (1.73%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22530 (98.16%) aligned exactly 1 time\n",
      "    422 (1.84%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22569 (98.33%) aligned exactly 1 time\n",
      "    383 (1.67%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22582 (98.39%) aligned exactly 1 time\n",
      "    370 (1.61%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22506 (98.06%) aligned exactly 1 time\n",
      "    446 (1.94%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22552 (98.26%) aligned exactly 1 time\n",
      "    400 (1.74%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22552 (98.26%) aligned exactly 1 time\n",
      "    400 (1.74%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22547 (98.24%) aligned exactly 1 time\n",
      "    405 (1.76%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22549 (98.24%) aligned exactly 1 time\n",
      "    403 (1.76%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22558 (98.28%) aligned exactly 1 time\n",
      "    394 (1.72%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22531 (98.17%) aligned exactly 1 time\n",
      "    421 (1.83%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22529 (98.16%) aligned exactly 1 time\n",
      "    423 (1.84%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22531 (98.17%) aligned exactly 1 time\n",
      "    421 (1.83%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22558 (98.28%) aligned exactly 1 time\n",
      "    394 (1.72%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22508 (98.07%) aligned exactly 1 time\n",
      "    444 (1.93%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22543 (98.22%) aligned exactly 1 time\n",
      "    409 (1.78%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22523 (98.13%) aligned exactly 1 time\n",
      "    429 (1.87%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22542 (98.21%) aligned exactly 1 time\n",
      "    410 (1.79%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22548 (98.24%) aligned exactly 1 time\n",
      "    404 (1.76%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22555 (98.27%) aligned exactly 1 time\n",
      "    397 (1.73%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22570 (98.34%) aligned exactly 1 time\n",
      "    382 (1.66%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22538 (98.20%) aligned exactly 1 time\n",
      "    414 (1.80%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22543 (98.22%) aligned exactly 1 time\n",
      "    409 (1.78%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22536 (98.19%) aligned exactly 1 time\n",
      "    416 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22535 (98.18%) aligned exactly 1 time\n",
      "    417 (1.82%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22562 (98.30%) aligned exactly 1 time\n",
      "    390 (1.70%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22516 (98.10%) aligned exactly 1 time\n",
      "    436 (1.90%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22461 (97.86%) aligned exactly 1 time\n",
      "    491 (2.14%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22496 (98.01%) aligned exactly 1 time\n",
      "    456 (1.99%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22574 (98.35%) aligned exactly 1 time\n",
      "    378 (1.65%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22576 (98.36%) aligned exactly 1 time\n",
      "    376 (1.64%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22559 (98.29%) aligned exactly 1 time\n",
      "    393 (1.71%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "22952 reads; of these:\n",
      "  22952 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    22536 (98.19%) aligned exactly 1 time\n",
      "    416 (1.81%) aligned >1 times\n",
      "100.00% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "# Align all of the DNA reads; this will align all the DNA reads. \n",
    "for i in range(1, 101):\n",
    "    !bowtie2 -f -x ../class_09/sample_genome -U data/DNA/sample_reads_P{i}.fa.gz -S data/DNA/sample_reads_P{i}.sam\n",
    "    !samtools sort -o data/DNA/sample_reads_P{i}.bam data/DNA/sample_reads_P{i}.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5755 reads; of these:\n",
      "  5755 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5747 (99.86%) aligned exactly 1 time\n",
      "    8 (0.14%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5741 reads; of these:\n",
      "  5741 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5731 (99.83%) aligned exactly 1 time\n",
      "    10 (0.17%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4431 reads; of these:\n",
      "  4431 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4426 (99.89%) aligned exactly 1 time\n",
      "    5 (0.11%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5372 reads; of these:\n",
      "  5372 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5370 (99.96%) aligned exactly 1 time\n",
      "    2 (0.04%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5008 reads; of these:\n",
      "  5008 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4996 (99.76%) aligned exactly 1 time\n",
      "    12 (0.24%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5562 reads; of these:\n",
      "  5562 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5551 (99.80%) aligned exactly 1 time\n",
      "    11 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5107 reads; of these:\n",
      "  5107 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5100 (99.86%) aligned exactly 1 time\n",
      "    7 (0.14%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4726 reads; of these:\n",
      "  4726 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4719 (99.85%) aligned exactly 1 time\n",
      "    7 (0.15%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5285 reads; of these:\n",
      "  5285 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5282 (99.94%) aligned exactly 1 time\n",
      "    3 (0.06%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4745 reads; of these:\n",
      "  4745 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4738 (99.85%) aligned exactly 1 time\n",
      "    7 (0.15%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4563 reads; of these:\n",
      "  4563 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4546 (99.63%) aligned exactly 1 time\n",
      "    17 (0.37%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4726 reads; of these:\n",
      "  4726 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4714 (99.75%) aligned exactly 1 time\n",
      "    12 (0.25%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5202 reads; of these:\n",
      "  5202 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5188 (99.73%) aligned exactly 1 time\n",
      "    14 (0.27%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4708 reads; of these:\n",
      "  4708 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4697 (99.77%) aligned exactly 1 time\n",
      "    11 (0.23%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5199 reads; of these:\n",
      "  5199 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5186 (99.75%) aligned exactly 1 time\n",
      "    13 (0.25%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5568 reads; of these:\n",
      "  5568 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5557 (99.80%) aligned exactly 1 time\n",
      "    11 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5176 reads; of these:\n",
      "  5176 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5159 (99.67%) aligned exactly 1 time\n",
      "    17 (0.33%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5187 reads; of these:\n",
      "  5187 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5175 (99.77%) aligned exactly 1 time\n",
      "    12 (0.23%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5096 reads; of these:\n",
      "  5096 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5089 (99.86%) aligned exactly 1 time\n",
      "    7 (0.14%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4756 reads; of these:\n",
      "  4756 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4745 (99.77%) aligned exactly 1 time\n",
      "    11 (0.23%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5527 reads; of these:\n",
      "  5527 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5515 (99.78%) aligned exactly 1 time\n",
      "    12 (0.22%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5826 reads; of these:\n",
      "  5826 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5814 (99.79%) aligned exactly 1 time\n",
      "    12 (0.21%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4918 reads; of these:\n",
      "  4918 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4904 (99.72%) aligned exactly 1 time\n",
      "    14 (0.28%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4902 reads; of these:\n",
      "  4902 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4892 (99.80%) aligned exactly 1 time\n",
      "    10 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5353 reads; of these:\n",
      "  5353 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5346 (99.87%) aligned exactly 1 time\n",
      "    7 (0.13%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5546 reads; of these:\n",
      "  5546 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5530 (99.71%) aligned exactly 1 time\n",
      "    16 (0.29%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4548 reads; of these:\n",
      "  4548 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4533 (99.67%) aligned exactly 1 time\n",
      "    15 (0.33%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5549 reads; of these:\n",
      "  5549 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5537 (99.78%) aligned exactly 1 time\n",
      "    12 (0.22%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4903 reads; of these:\n",
      "  4903 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4888 (99.69%) aligned exactly 1 time\n",
      "    15 (0.31%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5759 reads; of these:\n",
      "  5759 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5749 (99.83%) aligned exactly 1 time\n",
      "    10 (0.17%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4448 reads; of these:\n",
      "  4448 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4442 (99.87%) aligned exactly 1 time\n",
      "    6 (0.13%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4833 reads; of these:\n",
      "  4833 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4818 (99.69%) aligned exactly 1 time\n",
      "    15 (0.31%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "6204 reads; of these:\n",
      "  6204 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    6198 (99.90%) aligned exactly 1 time\n",
      "    6 (0.10%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4451 reads; of these:\n",
      "  4451 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4444 (99.84%) aligned exactly 1 time\n",
      "    7 (0.16%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4433 reads; of these:\n",
      "  4433 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4426 (99.84%) aligned exactly 1 time\n",
      "    7 (0.16%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5168 reads; of these:\n",
      "  5168 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5155 (99.75%) aligned exactly 1 time\n",
      "    13 (0.25%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5180 reads; of these:\n",
      "  5180 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5166 (99.73%) aligned exactly 1 time\n",
      "    14 (0.27%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5833 reads; of these:\n",
      "  5833 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5821 (99.79%) aligned exactly 1 time\n",
      "    12 (0.21%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "6222 reads; of these:\n",
      "  6222 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    6208 (99.77%) aligned exactly 1 time\n",
      "    14 (0.23%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5827 reads; of these:\n",
      "  5827 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5818 (99.85%) aligned exactly 1 time\n",
      "    9 (0.15%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4089 reads; of these:\n",
      "  4089 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4080 (99.78%) aligned exactly 1 time\n",
      "    9 (0.22%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5752 reads; of these:\n",
      "  5752 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5744 (99.86%) aligned exactly 1 time\n",
      "    8 (0.14%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5818 reads; of these:\n",
      "  5818 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5807 (99.81%) aligned exactly 1 time\n",
      "    11 (0.19%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5728 reads; of these:\n",
      "  5728 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5723 (99.91%) aligned exactly 1 time\n",
      "    5 (0.09%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5370 reads; of these:\n",
      "  5370 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5361 (99.83%) aligned exactly 1 time\n",
      "    9 (0.17%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4922 reads; of these:\n",
      "  4922 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4913 (99.82%) aligned exactly 1 time\n",
      "    9 (0.18%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5197 reads; of these:\n",
      "  5197 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5188 (99.83%) aligned exactly 1 time\n",
      "    9 (0.17%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5079 reads; of these:\n",
      "  5079 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5067 (99.76%) aligned exactly 1 time\n",
      "    12 (0.24%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4712 reads; of these:\n",
      "  4712 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4701 (99.77%) aligned exactly 1 time\n",
      "    11 (0.23%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4905 reads; of these:\n",
      "  4905 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4901 (99.92%) aligned exactly 1 time\n",
      "    4 (0.08%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "6199 reads; of these:\n",
      "  6199 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    6181 (99.71%) aligned exactly 1 time\n",
      "    18 (0.29%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5184 reads; of these:\n",
      "  5184 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5174 (99.81%) aligned exactly 1 time\n",
      "    10 (0.19%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5374 reads; of these:\n",
      "  5374 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5366 (99.85%) aligned exactly 1 time\n",
      "    8 (0.15%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5188 reads; of these:\n",
      "  5188 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5177 (99.79%) aligned exactly 1 time\n",
      "    11 (0.21%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5554 reads; of these:\n",
      "  5554 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5544 (99.82%) aligned exactly 1 time\n",
      "    10 (0.18%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4726 reads; of these:\n",
      "  4726 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4725 (99.98%) aligned exactly 1 time\n",
      "    1 (0.02%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5356 reads; of these:\n",
      "  5356 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5350 (99.89%) aligned exactly 1 time\n",
      "    6 (0.11%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4729 reads; of these:\n",
      "  4729 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4715 (99.70%) aligned exactly 1 time\n",
      "    14 (0.30%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4995 reads; of these:\n",
      "  4995 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4985 (99.80%) aligned exactly 1 time\n",
      "    10 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4453 reads; of these:\n",
      "  4453 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4443 (99.78%) aligned exactly 1 time\n",
      "    10 (0.22%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5850 reads; of these:\n",
      "  5850 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5834 (99.73%) aligned exactly 1 time\n",
      "    16 (0.27%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5821 reads; of these:\n",
      "  5821 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5810 (99.81%) aligned exactly 1 time\n",
      "    11 (0.19%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5384 reads; of these:\n",
      "  5384 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5376 (99.85%) aligned exactly 1 time\n",
      "    8 (0.15%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5378 reads; of these:\n",
      "  5378 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5369 (99.83%) aligned exactly 1 time\n",
      "    9 (0.17%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5091 reads; of these:\n",
      "  5091 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5081 (99.80%) aligned exactly 1 time\n",
      "    10 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4933 reads; of these:\n",
      "  4933 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4918 (99.70%) aligned exactly 1 time\n",
      "    15 (0.30%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5824 reads; of these:\n",
      "  5824 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5811 (99.78%) aligned exactly 1 time\n",
      "    13 (0.22%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4562 reads; of these:\n",
      "  4562 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4552 (99.78%) aligned exactly 1 time\n",
      "    10 (0.22%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5383 reads; of these:\n",
      "  5383 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5372 (99.80%) aligned exactly 1 time\n",
      "    11 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5113 reads; of these:\n",
      "  5113 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5106 (99.86%) aligned exactly 1 time\n",
      "    7 (0.14%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5193 reads; of these:\n",
      "  5193 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5184 (99.83%) aligned exactly 1 time\n",
      "    9 (0.17%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5187 reads; of these:\n",
      "  5187 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5173 (99.73%) aligned exactly 1 time\n",
      "    14 (0.27%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5579 reads; of these:\n",
      "  5579 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5570 (99.84%) aligned exactly 1 time\n",
      "    9 (0.16%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5370 reads; of these:\n",
      "  5370 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5365 (99.91%) aligned exactly 1 time\n",
      "    5 (0.09%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5184 reads; of these:\n",
      "  5184 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5172 (99.77%) aligned exactly 1 time\n",
      "    12 (0.23%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "6215 reads; of these:\n",
      "  6215 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    6203 (99.81%) aligned exactly 1 time\n",
      "    12 (0.19%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "6201 reads; of these:\n",
      "  6201 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    6186 (99.76%) aligned exactly 1 time\n",
      "    15 (0.24%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5366 reads; of these:\n",
      "  5366 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5355 (99.80%) aligned exactly 1 time\n",
      "    11 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4367 reads; of these:\n",
      "  4367 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4357 (99.77%) aligned exactly 1 time\n",
      "    10 (0.23%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5110 reads; of these:\n",
      "  5110 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5100 (99.80%) aligned exactly 1 time\n",
      "    10 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "6190 reads; of these:\n",
      "  6190 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    6181 (99.85%) aligned exactly 1 time\n",
      "    9 (0.15%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4721 reads; of these:\n",
      "  4721 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4713 (99.83%) aligned exactly 1 time\n",
      "    8 (0.17%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5548 reads; of these:\n",
      "  5548 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5536 (99.78%) aligned exactly 1 time\n",
      "    12 (0.22%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5115 reads; of these:\n",
      "  5115 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5107 (99.84%) aligned exactly 1 time\n",
      "    8 (0.16%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4845 reads; of these:\n",
      "  4845 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4831 (99.71%) aligned exactly 1 time\n",
      "    14 (0.29%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5372 reads; of these:\n",
      "  5372 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5362 (99.81%) aligned exactly 1 time\n",
      "    10 (0.19%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5095 reads; of these:\n",
      "  5095 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5085 (99.80%) aligned exactly 1 time\n",
      "    10 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "6222 reads; of these:\n",
      "  6222 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    6210 (99.81%) aligned exactly 1 time\n",
      "    12 (0.19%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5547 reads; of these:\n",
      "  5547 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5533 (99.75%) aligned exactly 1 time\n",
      "    14 (0.25%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5721 reads; of these:\n",
      "  5721 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5714 (99.88%) aligned exactly 1 time\n",
      "    7 (0.12%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5187 reads; of these:\n",
      "  5187 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5175 (99.77%) aligned exactly 1 time\n",
      "    12 (0.23%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4450 reads; of these:\n",
      "  4450 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4441 (99.80%) aligned exactly 1 time\n",
      "    9 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5100 reads; of these:\n",
      "  5100 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5094 (99.88%) aligned exactly 1 time\n",
      "    6 (0.12%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4085 reads; of these:\n",
      "  4085 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4080 (99.88%) aligned exactly 1 time\n",
      "    5 (0.12%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4241 reads; of these:\n",
      "  4241 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4235 (99.86%) aligned exactly 1 time\n",
      "    6 (0.14%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "5390 reads; of these:\n",
      "  5390 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    5382 (99.85%) aligned exactly 1 time\n",
      "    8 (0.15%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "6220 reads; of these:\n",
      "  6220 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    6207 (99.79%) aligned exactly 1 time\n",
      "    13 (0.21%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4468 reads; of these:\n",
      "  4468 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4459 (99.80%) aligned exactly 1 time\n",
      "    9 (0.20%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "4634 reads; of these:\n",
      "  4634 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    4631 (99.94%) aligned exactly 1 time\n",
      "    3 (0.06%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "6212 reads; of these:\n",
      "  6212 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    6201 (99.82%) aligned exactly 1 time\n",
      "    11 (0.18%) aligned >1 times\n",
      "100.00% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "# Process RNA the same way:\n",
    "for i in range(1, 101):\n",
    "    !bowtie2 -f -x ../class_09/sample_genome -U data/RNA/sample_RNA_reads_P{i}.fa.gz -S data/RNA/sample_RNA_reads_P{i}.sam\n",
    "    !samtools sort -o data/RNA/sample_RNA_reads_P{i}.bam data/RNA/sample_RNA_reads_P{i}.sam\n",
    "    !samtools index data/RNA/sample_RNA_reads_P{i}.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate the genotypes\n",
    "\n",
    "Next, we can calculate the genotypes for all of the individuals. We did this two classes ago and now just need to keep track of it for each person. Our end goal of this step is mostly just reformatting the data (this is very common in bioinformatics). We would like to have an array of variants by samples that contains the genotype of a sample for each variant. To build a master variant list, we will be using the criteria of common variants (>1% population frequency). This reduces the number of tests that we will have to perform.\n",
    "\n",
    "The first step of this will be to collate all of the variants into a list:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from genotyper import *\n",
    "\n",
    "#calculate the genotypes for all the individuals, and now we need to keep track of it for each individual. \n",
    "#this function will \n",
    "\n",
    "def get_all_variants():\n",
    "    ''' Function to iterate through all DNA files to get genotypes\n",
    "    \n",
    "    This functions runs our previous find_variants() function on all of the individuals and makes a list of these\n",
    "    \n",
    "    Example:\n",
    "        >>> get_all_variants() #doctest: +ELLIPSIS\n",
    "        [(803, 'C', 'A'), (5144, 'G', 'C'), (5275, 'T', 'T'), ...\n",
    "    '''\n",
    "    individual_variants = []\n",
    "\n",
    "    # Note that this now starts at index 0\n",
    "    for i in range(1, 101):\n",
    "        # Run find_variants and store in a list indexed by sample\n",
    "        individual_variants.append(find_variants(reference = '../class_09/data/sample_genome.fna', alignments = 'data/DNA/sample_reads_P' + str(i) + '.bam'))\n",
    "\n",
    "    return individual_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " UserWarning:/Users/alauth/opt/anaconda3/envs/b529/lib/python3.7/site-packages/bamnostic/bam.py:310: No supplied index file was not found. Random access disabled\n"
     ]
    }
   ],
   "source": [
    "# This takes a few minutes to run\n",
    "#this sets it to a variable name. \n",
    "individual_genotypes = get_all_variants()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(individual_genotypes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next step will be to first identify our final variant list (`build_list_of_common_variants`) and then use that list along with the list of genotypes to build our final bulk matrix described above (`build_bulk_genotype`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "import numpy as np\n",
    "\n",
    "def build_list_of_common_variants(individual_genotypes):\n",
    "    ''' This function will combine all of the genotypes to create a master list of SNP locations\n",
    "        for common variants. Common variants are those that appear at a frequency of more than 1%\n",
    "        in the population.\n",
    "        \n",
    "        Note: we only need to track the position of the SNPs\n",
    "    '''\n",
    "    \n",
    "    all_variant_counter = Counter()\n",
    "    all_variant_list = []\n",
    "    #which variants are the most common aka they are above 1% threshold statistically \n",
    "    \n",
    "    \n",
    "    # Here I use a Counter() to just count the frequency of each variant location across the samples\n",
    "    for i, sample in enumerate(individual_genotypes):\n",
    "        for j, variant in enumerate(sample):\n",
    "            position, allele1, allele2 = variant\n",
    "            all_variant_counter[position] += 1\n",
    "            \n",
    "            print(all_variant_counter)\n",
    "    \n",
    "    # Now we create a list from just the counters that are above 1% (or 1 since we have 100 samples)\n",
    "    for key in list(all_variant_counter):\n",
    "        if all_variant_counter[key] > 1:\n",
    "            all_variant_list.append(key)\n",
    "            \n",
    "            # print(key)\n",
    "            \n",
    "    return all_variant_list\n",
    "    \n",
    "def build_bulk_genotype(all_variants, individual_genotypes):\n",
    "    ''' This function will assign one of 3 genotypes at every SNP location\n",
    "    ref/ref = 0  check and see if a particular position was called in a sample. \n",
    "    ref/alt = 1\n",
    "    alt/alt = 2\n",
    "    \n",
    "    This will populate a varaint x individual matrix with the 2 genotypes maintaining individual order from \n",
    "    individual_genotypes and variant order from all_variants\n",
    "    \n",
    "    Pseudocode:\n",
    "    For each variant\n",
    "        For each individual\n",
    "            if individual has variant assign value 1 if het, 2 if homozygous alt\n",
    "            otherwise value is 0\n",
    "    \n",
    "    Args:\n",
    "        all_variants (list): list of common variants\n",
    "        individual_genotypes (list of list of tuples): list of all variants for each individual\n",
    "        \n",
    "    Returns:\n",
    "        sample_variant_matrix (numpy int matrix): matrix of variants x individuals\n",
    " \n",
    "    '''\n",
    "    \n",
    "    #all_variants (SNPs with just positions) so this SNP was present in at least 2 people in this 100 population group \n",
    "    #individl contains the position, ref all, alt all, \n",
    "    #index indivual_genotypes with positino from all variants, then compare alleles \n",
    "    #then plug and chug into the array. \n",
    "    #if ref/ref need to iterate through rest of pop and if they don't have it, put a \"-\" \n",
    "    \n",
    "    #initialize the empty matrix here:\n",
    "    sample_variant_matrix = np.zeros((len(all_variants), len(individual_genotpyes)), dtype=int)\n",
    "    \n",
    "    \n",
    "    #for each variant\n",
    "    for index, variant_pos in enumerate(all_variants):\n",
    "        #for loop through each individual\n",
    "        for index2, variants in enumerate(individual_genotypes):\n",
    "            #if we have not seen the variant it is zero (homozygous) \n",
    "            #our initial (if its ref/ref)\n",
    "            sample_genotype = 0\n",
    "            \n",
    "            #all the individual genotypes for each individual. \n",
    "            #must remember that each individual has it's own individual genotypes too, so you need to index for that too. \n",
    "            #that is why you need to index further. \n",
    "            for index3, variant in enumerate(individual_genotypes[index2]):\n",
    "                pos, allele1, allele2 = variant\n",
    "                if pos == variant_pos:\n",
    "                    if allele1 == allele2:\n",
    "                        #homozygous alternative \n",
    "                        sample_genotype = 2\n",
    "                    if allele1 != allele2:\n",
    "                        #heterozygous \n",
    "                        sample_genotype = 1\n",
    "                    break \n",
    "                \n",
    "            #need the position that is index of the all_variants and the individual_genotypes. \n",
    "            sample_variant_matrix[index][index2] = sample_genotype  \n",
    "    \n",
    "    \n",
    "    return sample_variant_matrix\n",
    "                \n",
    "    #now need to iterate through the variant list (call the helper funtion and now the individual_genotypes. \n",
    "    # Pseudocode:\n",
    "    # For each variant\n",
    "    #     For each individual\n",
    "    #         if individual has variant assign value 1 if het, 2 if homozygous alt\n",
    "    #         otherwise value is 0\n",
    "    \n",
    "    \n",
    "    #reference allele if you don't have a record for that position, then that's the reference. \n",
    "    \n",
    "#take genotypes and associate them with gene expression in the samples \n",
    "#doing a linear model, how genotype as a category is associated with a continuous variable in the RPKMs. \n",
    "#we care about the category (homo, or hetero) \n",
    "    \n",
    "#go through each of the tuples and check if they are \n",
    "#aggregate list of all the positions, "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # # Now we get the final bulk_genotype matrix\n",
    "# full_variant_list = build_list_of_common_variants(individual_genotypes)\n",
    "# bulk_genotypes = build_bulk_genotype(full_variant_list, individual_genotypes)\n",
    "# print(bulk_genotypes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate the gene expression\n",
    "\n",
    "In addition to the genotype matrix above, we also need to generate expression data for all of the individuals at each gene. To do this, we can use the methods from the previous class `get_transcript_levels` to get the expression level of each gene for each person. As above, we will store this as a matrix of gene by samples with RPKM in each position of the matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rna_tools import *\n",
    "\n",
    "def get_all_expression():\n",
    "    ''' Function to iterate through all RNA file to get expression levels of genes\n",
    "    '''\n",
    "    individual_expression_matrix = np.zeros((5,100), dtype=float)\n",
    "\n",
    "    # Note that this now starts at index 0\n",
    "    for i in range(1, 101):\n",
    "        # Run our get_transcript_levels for all individuals\n",
    "        individual_expression = get_transcript_levels('data/RNA/sample_RNA_reads_P' + str(i) + '.bam', '../class_10/data/sample_genomic.gff')\n",
    "        for j, data in enumerate(individual_expression):\n",
    "            seqid, start, end, expression = data\n",
    "            individual_expression_matrix[j][i-1] = expression # Just save the expression value\n",
    "        \n",
    "    return individual_expression_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This takes a few minutes to run\n",
    "bulk_expression = get_all_expression()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## eQTL Analysis\n",
    "\n",
    "As discussed in the class materials, the variants that we have identified from our diploid samples represent polymorphic locations in the sample genomes. These sites can sometimes affect the regulation or expression level of a gene, often through disruption of transcription factor binding. Identifying variants associated with this change in expression is thus of great interest in understanding how the gene expression might be controlled. To do this, we use eQTL analysis.\n",
    "\n",
    "Standard eQTL analysis involves direct pairwise testing between genetic markers (here SNPs) andd the gene expression levels across many (here 100) individuals. This can be done for all SNPs vs all genes, but this results in a large penalty when accounting for multiple testing. To avoid this, the analysis is often done in a window around each gene of ~ 1 magabase. For our example today, we can just test all of the variants because our genome is quite small.\n",
    "\n",
    "Typical eQTL analysis uses a generalized linear model (GLM). For simplicity, we will use a simple linear model of the form $Y_{i} = b_{0} + b_{1}X_{i} + \\epsilon_{i}$ where $Y_{i}$ is the RPKM for individual $i$ and $X_{i}$ indicates the genotype (0, 1, 2) at a given SNP. We can implement this with the `stats.linregress()` function in scipy ( https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.linregress.html ).\n",
    "\n",
    "We can also plot these associations using `matplotlib.pyplot` ( https://matplotlib.org/api/_as_gen/matplotlib.pyplot.html ).\n",
    "\n",
    "<img src = \"figures/eQTL_fig.png\">\n",
    "\n",
    "Today, generate the variant and gene associations that are significant ($p < .01$) and plot these below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def eQTL_analysis(bulk_genotypes, bulk_expression):\n",
    "    ''' Function to perform eQTL analysis\n",
    "    \n",
    "    Args:\n",
    "        bulk_genotypes (2D np array): genotypes for each variant for each individual\n",
    "        bulk_expression (2D np array): expression values for each gene for each individual\n",
    "       \n",
    "    Prints:\n",
    "        significant eQTLs (p < .01) and plots them as above\n",
    "    '''\n",
    "    \n",
    "    for gene_index in range(bulk_expression.shape[0]):\n",
    "        y = np.array(bulk_expression[gene_index]) #give us all the values from the gene_index. \n",
    "        \n",
    "        for snp_index in range(bulk_genotypes.shape[0]):\n",
    "            xi = np.array(bulk_genotypes[snp_index])\n",
    "            slope, intercept, r_value, p_value, std_err = stats.lineregress(xi, y)\n",
    "            \n",
    "            if p_value < 0.1:\n",
    "                print(\"significant eQTL: \", \"gene\", gene_index, \"snp\", snp_index, \"p_value\", p_value)\n",
    "                line = slope * xi * intercept \n",
    "                \n",
    "                plt.plot(xi, line, \"r-\", xi, y, \"o\")\n",
    "                plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "eQTL_analysis(bulk_genotypes, bulk_expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
